### PR TITLE
Merge 2.9

### PR DESCRIPTION
**Pull request by wallyworld**
Merge 2.9

#14220 Fix k8s test race
#14223 [JUJU-1382] Improve formatting of machine errors in state
#14183 [JUJU-1340] Deprecate win and mongo older than 4.4
#14227 [JUJU-1384] Use --client for add-k8s command;
#14226 Add a new model upgrader facade and update upgrade command to use it

The main conflicts were in the upgrade command and tests.
```
# Conflicts:
#       api/client/modelmanager/modelmanager.go
#       api/client/modelmanager/modelmanager_test.go
#       apiserver/facades/client/client/client.go
#       apiserver/facades/client/client/client_test.go
#       apiserver/facades/client/modelmanager/modelinfo_test.go
#       apiserver/facades/client/modelmanager/modelmanager.go
#       apiserver/facades/client/modelmanager/modelmanager_test.go
#       cmd/juju/commands/upgrademodel.go
#       cmd/juju/commands/upgrademodel_test.go
#       core/series/supportedseries.go
#       migration/precheck_test.go
#       upgrades/model.go
#       upgrades/model_test.go
```

## QA steps

See PRs

[JUJU-1382]: https://warthogs.atlassian.net/browse/JUJU-1382?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1340]: https://warthogs.atlassian.net/browse/JUJU-1340?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[JUJU-1384]: https://warthogs.atlassian.net/browse/JUJU-1384?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by jameinel — APPROVED** [review]
mostly rubber stamping a forward merge

File: cmd/juju/commands/upgrademodel.go [hunk 1]
```diff
@@ -186,7 +186,8 @@ func (c *baseUpgradeCommand) precheckVersion(ctx *cmd.Context, agentVersion vers
 		// Only upgrade to a different major number if:
 		// 1 - Explicitly requested with --agent-version or using --build-agent, and
 		// 2 - The model is running a valid version to upgrade from.
-		allowed, minVer, err := upgrades.UpgradeAllowed(agentVersion, c.Version)
+		// We will do this check in server side later, but it doesn't hurt to do it here as a precheck.
+		allowed, minVer, err := upgradevalidation.UpgradeToAllowed(agentVersion, c.Version)
```
Comment (ycliuhw) [inline cmd/juju/commands/upgrademodel.go:190]: So it seems we forgot to remove this.

**Review by ycliuhw — COMMENTED** [review]
Thank you for getting this merge!

**Pull request merged**